File: src/0018_4sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[][]}
 */
var fourSum = function (nums, target) {
    nums.sort((a, b) => a - b);
    const results = [];

    const findNsum = (l, r, target, n, result) => {
        if (r - l + 1 < n || n < 2 || target < nums[l] * n || target > nums[r] * n) {
            return;
        }
        if (n === 2) {
            while (l < r) {
                const s = nums[l] + nums[r];
                if (s === target) {
                    results.push(result.concat([nums[l], nums[r]]));
                    l += 1;
                    while (l < r && nums[l] === nums[l - 1]) {
                        l += 1;
                    }
                } else if (s < target) {
                    l += 1;
                } else {
                    r -= 1;
                }
            }
        } else {
            for (let i = l; i <= r; i++) {
                if (i === l || (i > l && nums[i - 1] !== nums[i])) {
                    findNsum(i + 1, r, target - nums[i], n - 1, result.concat([nums[i]]));
                }
            }
        }
    };
    findNsum(0, nums.length - 1, target, 4, []);
    return results;
};

module.exports = { fourSum };

if (require.main === module) {
    const assert = require('assert');
    function ListNode(val = 0, next = null) { this.val = val; this.next = next; }
    function TreeNode(val = 0, left = null, right = null) { this.val = val; this.left = left; this.right = right; }
    function Node(val = 0, neighbors = []) { this.val = val; this.neighbors = neighbors; }
    function listNodeToString(node) {
        const values = [];
        let current = node;
        while (current !== null) {
            values.push(String(current.val));
            current = current.next;
        }
        return values.join('->');
    }
    function treeToString(node) {
        if (node === null) return 'None';
        return `${node.val}(${treeToString(node.left)},${treeToString(node.right)})`;
    }
    function sorted(value) { return value.slice().sort(); }
    assert.deepStrictEqual(fourSum([1, 0, (-1), 0, (-2), 2], 0), [[(-2), (-1), 1, 2], [(-2), 0, 0, 2], [(-1), 0, 0, 1]]);
}

File: src/0018_4sum.rs
struct Solution;

impl Solution {
    pub fn four_sum(mut nums: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
        nums.sort();
        let mut result = Vec::new();
        let n = nums.len();
        for i in 0..n {
            if i > 0 && nums[i] == nums[i - 1] {
                continue;
            }
            for j in i + 1..n {
                if j > i + 1 && nums[j] == nums[j - 1] {
                    continue;
                }
                let mut left = j + 1;
                let mut right = n.saturating_sub(1);
                while left < right {
                    let total =
                        nums[i] as i64 + nums[j] as i64 + nums[left] as i64 + nums[right] as i64;
                    if total == target as i64 {
                        result.push(vec![nums[i], nums[j], nums[left], nums[right]]);
                        left += 1;
                        right -= 1;
                        while left < right && nums[left] == nums[left - 1] {
                            left += 1;
                        }
                        while left < right && nums[right] == nums[right + 1] {
                            right -= 1;
                        }
                    } else if total < target as i64 {
                        left += 1;
                    } else {
                        right -= 1;
                    }
                }
            }
        }
        result
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn main_asserts() {
        assert_eq!(
            Solution::four_sum(vec![1, 0, -1, 0, -2, 2], 0),
            vec![vec![-2, -1, 1, 2], vec![-2, 0, 0, 2], vec![-1, 0, 0, 1]]
        );
    }
}
